cpp/src/arrow/array/data.cc

```diff
@@ -228,11 +228,12 @@ void ArraySpan::SetMembers(const ArrayData& data) {
 namespace {
 
 template <typename offset_type>
-void SetOffsetsForScalar(ArraySpan* span, offset_type* buffer, int64_t value_size,
+void SetOffsetsForScalar(ArraySpan* span, uint8_t* buffer, int64_t value_size,
```

Comment:
`value_size` could be declared as `offset_type` instead of `int64_t`.

Reply:
We tend to use `int64_t` for sizes consistently, so let's continue to do that here, IMHO.

Reply:
I think the cast at the callsite is the right way to go because by looking at just `SetOffsetsForScalar` I can't say it never truncates values into a 32-bit `offset_type`.

@pitrou but this is an offset buffer, not a single array length/offset.